src/core/ComboBox/ComboBox.tsx

```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+          key={option.id ?? __originalIndex}
+          {...rest}
+          isSelected={selectedIndex === __originalIndex}
+          onClick={() => dispatch(['select', __originalIndex])}
+          index={__originalIndex}
+        >
+          {option.label}
+        </ComboBoxMenuItem>
       );
-      switch (event.key) {
-        case 'ArrowDown':
-          if (isOpen) {
-            const nextIndex = Math.min(
-              focusedIndexInFilteredList + 1,
-              focusableOptions.length - 1,
-            );
-            setFocusedIndex(
-              options.findIndex(
-                (_, index) =>
-                  getOptionId(index) === focusableOptions[nextIndex].id,
-              ),
-            );
-          } else {
-            setIsOpen(true); // reopen menu if closed when typing
-          }
-          event.preventDefault();
-          event.stopPropagation();
-          break;
-        case 'ArrowUp':
-          if (isOpen) {
-            const previousIndex = Math.max(focusedIndexInFilteredList - 1, 0);
-            setFocusedIndex(
-              options.findIndex(
-                (_, index) =>
-                  getOptionId(index) === focusableOptions[previousIndex].id,
-              ),
-            );
-          }
-          event.preventDefault();
-          event.stopPropagation();
-          break;
-        case 'Enter':
-          if (isOpen) {
-            setSelectedValue(options[focusedIndex].value);
-            userOnChange.current?.(options[focusedIndex].value);
-          }
-          setIsOpen((open) => !open);
-          event.preventDefault();
-          event.stopPropagation();
-          break;
-        case 'Escape':
-          setIsOpen(false);
-          event.preventDefault();
-          event.stopPropagation();
-          break;
-        case 'Tab':
-          setIsOpen(false);
-          break;
-        default:
-          if (!isOpen) {
-            setIsOpen(true); // reopen menu if closed when typing
-          }
-          break;
-      }
     },
-    [focusedIndex, isOpen, options, getOptionId],
+    [focusedIndex, itemRenderer, memoizedItems, selectedIndex],
   );
 
-  const menuItems = React.useMemo(() => {
-    if (filteredOptions.length === 0) {
-      return [
-        <MenuExtraContent key={0}>
-          <Text isMuted>{emptyStateMessage}</Text>
-        </MenuExtraContent>,
-      ];
+  return (
+    <ComboBoxRefsContext.Provider
+      value={{ inputRef, menuRef, toggleButtonRef, optionsRef }}
+    >
+      <ComboBoxActionContext.Provider value={dispatch}>
+        <ComboBoxStateContext.Provider
+          value={{ id, minWidth, isOpen, focusedIndex }}
+        >
+          {children ?? (
+            <>
+              <ComboBoxInputContainer
+                disabled={inputProps?.disabled}
+                endIcon={
+                  <ComboBoxEndIcon
+                    className={cx({
+                      'iui-actionable': !inputProps?.disabled,
```

Comment:
These could be props and classes could be moved inside icon component.

Reply:
So you want 3 new props on this super simple component?

Reply:
Added